ubirch/crypto.go

```diff
@@ -189,20 +195,71 @@ func (c *CryptoContext) SetKey(name string, id uuid.UUID, privKeyBytes []byte) e
 // Get a certificate signing request.
 func (c *CryptoContext) GetCSR(name string) ([]byte, error) { return nil, nil }
 
+// Get the decoded public key for the given name.
+func (c *CryptoContext) getDecodedPublicKey(name string) (*ecdsa.PublicKey, error) {
+	id, err := c.GetUUID(name)
+	if err != nil {
+		return nil, err
+	}
+
+	pubKey, err := c.Keystore.GetKey(pubKeyEntryTitle(id))
+	if err != nil {
+		return nil, err
+	}
+
+	// decode the key
+	return decodePublicKey(pubKey)
+}
+
 // Get the public key bytes for the given name.
 func (c *CryptoContext) GetPublicKey(name string) ([]byte, error) {
+	decodedPubKey, err := c.getDecodedPublicKey(name)
+	if err != nil {
+		return nil, fmt.Errorf("decoding public key from keystore failed: %s", err)
+	}
+
+	pubKeyBytes := make([]byte, 0, 0)
+
+	paddedX := make([]byte, 32)
+	paddedY := make([]byte, 32)
+	copy(paddedX[32-len(decodedPubKey.X.Bytes()):], decodedPubKey.X.Bytes())
+	copy(paddedY[32-len(decodedPubKey.Y.Bytes()):], decodedPubKey.Y.Bytes())
+	pubKeyBytes = append(pubKeyBytes, paddedX...)
+	pubKeyBytes = append(pubKeyBytes, paddedY...)
+
+	return pubKeyBytes, nil
+}
+
+// Get the decoded private key for the given name.
+func (c *CryptoContext) getDecodedPrivateKey(name string) (*ecdsa.PrivateKey, error) {
 	id, err := c.GetUUID(name)
 	if err != nil {
 		return nil, err
 	}
 
-	pubKeyBytes, err := c.Keystore.GetKey(pubKeyEntryTitle(id))
+	privKey, err := c.Keystore.GetKey(privKeyEntryTitle(id))
 	if err != nil {
 		return nil, err
 	}
-	return pubKeyBytes, nil
+
+	// decode the key
+	return decodePrivateKey(privKey)
+}
+
+// Check if a private key entry for the given name exists in the keystore.
+func (c *CryptoContext) PrivateKeyExists(name string) bool {
+	_, err := c.getDecodedPrivateKey(name)
+	if err != nil {
+		return false
+	}
+	return true
```

Comment:
This currently returns false even if the problem is not a missing key but for example a wrong secret for the keystore. For the future would be nice to distinguish between errors and 'no key available'. Maybe for now simply add a todo here: "Change to return bool, err, as soon as the keystore supports checking for key existence directly to distinguish between general errors and key really not avilable."

Reply:
i think it is acceptable to return false if there is no valid key in the keystore

Reply:
I like it better if I am able to distinguish (from the 'user side') between  'checking worked but there is no key', 'checking worked and there is a key' and 'something went wrong and thus I cant tell you if there is key'. But I guess this is fine for now since the keystore is missing suitable check functions.